### ISSUE TITLE
Add support for ```stablehlo.optimization_barrier```

### ISSUE THREAD

**Issue by umalesTT** [tenstorrent/tt-mlir#4034]
### Describe the bug

For training models at scale, two very important strategies are [FSDP](https://engineering.fb.com/2021/07/15/open-source/fsdp/) and [Gradient checkpointing](https://docs.jax.dev/en/latest/gradient-checkpointing.html). 

Both enforce recomputation during backward pass instead of caching values. JAX uses [remat policy](https://docs.jax.dev/en/latest/notebooks/autodiff_remat.html) to control this, and StableHLO uses [optimization barriers](https://openxla.org/stablehlo/spec#optimization_barrier)  to preserve these constraints when lowered from JAX.

Below is the example of usage of ```stablehlo.optimization_barrier``` 

```
// FORWARD PASS - creates operands for optimization_barrier:
%0 = stablehlo.dot_general %arg4, %arg0, contracting_dims = [1] x [0] # x * W1  
%1 = stablehlo.broadcast_in_dim %arg1, dims = [1] # broadcast b1
%2 = stablehlo.broadcast_in_dim %1, dims = [0, 1] # broadcast to batch
%3 = stablehlo.add %0, %2 # x * W1 + b1
%4 = call @relu(%3) # ReLU activation (stored as %4, not passed to barrier)

// BACKWARD PASS - recomputation after optimization_barrier:
%9:4 = stablehlo.optimization_barrier %arg0, %arg1, %arg2, %8 # barrier with x, W1, b1, upstream_grad
# After the barrier, the forward pass is recomputed:
%10 = stablehlo.dot_general %9#0, %9#1, contracting_dims = [1] x [0] # x * W1
%11 = stablehlo.broadcast_in_dim %9#2, dims = [1] # broadcast b1  
%12 = stablehlo.broadcast_in_dim %11, dims = [0, 1] # broadcast to batch
%13 = stablehlo.add %10, %12 # x * W1 + b1
# ReLU derivative computation:
%15 = stablehlo.compare GT, %13, %14, FLOAT # check where activations > 0
%17 = stablehlo.select %15, %9#3, %16 # apply ReLU derivative
```

The optimization barrier results may be viewed as if they came from an unknown graph, forcing everything depending on them to be recomputed. In this example, the expensive ReLU activation ```%4``` is not stored or passed through the barrier. Instead, only the raw parameters ```(x, W1, b1)``` and upstream gradient are preserved. During backward pass, the computation ```x * W1 + b1``` is recomputed from scratch. This trades memory (not storing activations) for computation (recomputing during backward pass), and the optimization barrier prevents compilers from optimizing away this intended recomputation pattern.

Official documentation https://openxla.org/stablehlo/spec#optimization_barrier

fyi @vladimirjovanovicTT @ajakovljevicTT @azecevicTT
